File: subsets/subsets.py
class Solution:
    def subsets(self, nums: List[int]) -> List[List[int]]:
        subsets = []

        def dfs(arr: List[int], nums: List[int]):
            subsets.append(arr)
            if not nums:
                return
            
            for i in range(len(nums)):
                if not arr or arr[-1] < nums[i]:
                    dfs(arr + [nums[i]], nums[:i] + nums[i+1:])

        dfs([], nums)
        return subsets

File: subsets/subsets.java
class Solution {
    public List<List<Integer>> subsets(int[] nums) {
        List<List<Integer>> result = new ArrayList<>();
        findSubset(nums, result, new ArrayList<>(), 0, nums.length);
        return result;
    }

    private void findSubset(int[] nums, List<List<Integer>> result, List<Integer> route, int start, int end) {
        result.add(new ArrayList<>(route));
        if (route.size() == nums.length) return;

        for (int i = start; i < end; i++) {
            route.add(nums[i]);
            findSubset(nums, result, route, i + 1, end);
            route.remove(route.size() - 1);
        }
    }
}